backend/fms_core/template_importer/sheet_data.py

```diff
@@ -0,0 +1,66 @@
+from django.core.exceptions import ValidationError
+from ._utils import data_row_ids_range, panda_values_to_str_list
+
+'''
+    SheetData objects
+    attributes (input): 
+        name, pandas dataframe, header row number
+
+    preview info from rows results (output): 
+        a dictionary with the sheet name, list of column headers, data sheet validity, 
+                              list of base_errors, list of rows_results
+'''
+
+
+class SheetData():
+    def __init__(self, name, dataframe, header_row_nb):
+        self.name = name
+        self.dataframe = dataframe
+        self.header_row_nb = header_row_nb
+
+        self.dataframe.columns = self.dataframe.values[self.header_row_nb]
+        print('SheetData columns: ', self.dataframe.columns)
+
+        self.base_errors = []
+        self.is_valid = None
+
+        self.prepare_rows()
+
+    def prepare_rows(self):
+        print('SheetData - start preparing rows')
+        self.rows = []
+        self.rows_results = []
+        for row_id in data_row_ids_range(self.header_row_nb + 1, self.dataframe):
+            row_data = self.dataframe.iloc[row_id]
+            self.rows.append(row_data)
+
+            print('row_data', row_data)
+
+            row_repr = f"#{row_id + 2}"
+
+            result = {
+                'row_repr': row_repr,
+                'diff': [row_repr] + panda_values_to_str_list(row_data),
```

Comment:
Maybe we could return the info as value before and after and do the diff and the repr stuff on the front-end side ?

Reply:
There is no 'before' and 'after' anymore. It is simply a copy of what was entered by the user in the Excel spreadsheet. I could rename 'diff' to a more appropriate name to the current context, such as 'cell_values' if preferred

Reply:
Then we need to add one ... simply printing the new value is a move in the wrong direction we are removing information users have at the moment. Users want to be able to tell from the confirmation table that the change they are submitting is correct. Simply posting back the template content will not help them do that.

Reply:
Maybe we can use the information stored in the front-end by using the ID of the object and compare them while rendering the report to prevent hitting the database anymore than needed. 
Some objects may not need before image also since they are always created and never updated.

Reply:
We have to verify at what extent the users use that. From what I recall this feature was half working with import-export and wasn't working at all for cases where the objects didnt follow closely the resource object.

This would require substantial work, and was not part of the technical requirements I initially listed.

Reply:
> Maybe we can use the information stored in the front-end by using the ID of the object and compare them while rendering the report to prevent hitting the database anymore than needed. Some objects may not need before image also since they are always created and never updated.

I don't agree with that solution. The diff, if any, would have to be done entirely in the back-end.

PS: The document 'Workflow Import Tool Freezeman' in 'Technical Analysis' in the FreezeMan shared drive is what I based my design on. If we are to state that this is an essential feature/a blocker, more work will have to be done.

Reply:
Without a comparison with the values that were present before the table is a useless copy of the template content. This would be sufficient for object creations (container, sample, individual, ...) but would be lacking for action like update and transfer. I agree that it is significant work ... but we spent significant time doing it once i think it warrant the need to do it again. Maybe we can reuse what is there now until we find the time ...

Reply:
To be discussed in team: at what point this is used by the lab. 
from my understanding it is a 'nice to have' but definitely not a blocker.

Reply:
Yeah lets discuss this in team...